torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp

```diff
@@ -398,19 +411,63 @@ std::shared_ptr<ReductionParams> innerPersistentHeuristic(
       : bdimx + (device_warp_size - bdimx % device_warp_size);
 
   bool pad_bdimx = bdimx > 16 &&
-      padded_bdimx * bdimy * bdimz <
-          (int64_t)at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock;
+      padded_bdimx * bdimy * bdimz < (int64_t)dev_prop->maxThreadsPerBlock;
 
   pad_bdimx = pad_bdimx &&
       bdimx * inner_reduction_unroll_factor != inner_most_dimension_numel;
 
+  int64_t nvrtc_register_per_thread = LaunchParams::UNINITIALIZED_VAL;
+  const int blocksPerKernel = godim;
+  if (vectorize && blocksPerKernel > device_multiprocessor_count &&
+      batches_per_block_inner_reduction > 1) {
+    // estimate register usage and occupancy raito.
+    // If occupancy raito is less than occupancy_ratio, reduce register usage
+    constexpr int reg_allocation_granularity = 256;
+    constexpr float occupancy_ratio = 0.4;
+    const int persistent_buffer_size = batches_per_block_inner_reduction *
+        inner_reduction_unroll_factor * max_input_dtype_size;
+    // persistent_buffer_size = 4*2, 8*2, 32*2, 64*2, 128*2
+    // register_used_on_a100  = 27,  40,  62,   73,   105
+    // register_used_on_v100  = xx,  xx,  45,   62,   93
+    // estimated_register_num = 42,  44,  56,   72,   104
+    // safe for both v100 & a100
+    const int estimated_register_count = persistent_buffer_size / 4 + 40;
+    const int register_per_warp =
+        ceilDiv(
+            estimated_register_count * device_warp_size,
+            reg_allocation_granularity) *
+        reg_allocation_granularity;
+    const int threadsPerBlock =
+        (pad_bdimx ? padded_bdimx : bdimx) * bdimy * bdimz;
+    const int warps_per_block = ceilDiv(threadsPerBlock, dev_prop->warpSize);
+    const int estimated_warps_per_sm = dev_prop->regsPerMultiprocessor /
+        (register_per_warp * warps_per_block) * warps_per_block;
+    const int occupancy_warps_per_sm = static_cast<int>(
+        dev_prop->maxThreadsPerMultiProcessor / device_warp_size *
+        occupancy_ratio);
+    if (estimated_warps_per_sm < occupancy_warps_per_sm) {
```

Comment:
I think maxrregcount should be set even if the estimated occupancy is higher than the target. Otherwise, nvcc is free to use more registers than the estimate.

Reply:
Added
I was thinking if there is no limitation on registers, nvcc may do a better job and shouldn't use too much more than our estimation and lead to occupancy issue.

Reply:
Which part did you change?

Reply:
// avoid nvcc using too many registers than expected
    nvrtc_register_per_thread = estimated_register_count;